### PR TITLE
`validate_call` type params fix

### PR DESCRIPTION
**Pull request by sydney-runkle**
Closes https://github.com/pydantic/pydantic/issues/9536
Fix https://github.com/pydantic/pydantic/issues/9744

### CODE REVIEW

File: tests/test_validate_call.py [hunk 1]
```diff
@@ -803,3 +804,44 @@ def foo(bar: 'list[int | str]') -> 'list[int | str]':
...
+from pydantic import validate_call
+
+@validate_call
+def max[T](args: Iterable[T]) -> T:
+    return sorted(args, reverse=True)[0]
+        """,
+        globs,
+    )
+    max = globs['max']
+    assert len(max.__type_params__) == 1
```
Comment (alexmojaki) [inline tests/test_validate_call.py:824]: Is there any way to test that `validate_call` is actually using/seeing `T`?

File: tests/test_validate_call.py [hunk 2]
```diff
@@ -803,3 +804,44 @@ def foo(bar: 'list[int | str]') -> 'list[int | str]':
...
+
+    with pytest.raises(ValidationError):
+        max(1)
+
+
+def test_uses_local_ns():
+    class M1(BaseModel):
+        x: int
+
+    M = M1  # noqa: F841
```
Comment (alexmojaki) [inline tests/test_validate_call.py:835]: please also add an `M` at the global level

File: tests/test_validate_call.py (hunk 1 above)
Comment (sydney-runkle) [reply tests/test_validate_call.py:824]: Well, I don't think it is: https://github.com/pydantic/pydantic/issues/7796

Which I think is something we can fix separately with the issue above

Comment (alexmojaki) [reply tests/test_validate_call.py:824]: but is there a type param in the pydantic schema or something? or is it replaced with `Any`?

Comment (sydney-runkle) [reply tests/test_validate_call.py]: It has an arguments schema with an `Any` for the item type

Comment (sydney-runkle) [reply tests/test_validate_call.py]: In a sense this PR is a bit silly, bc we're fixing the syntax for something that doesn't work, but it at least sets the stage to get the type params in accessible places.

**Review by alexmojaki — APPROVED** [review]

**Pull request merged**